Arm/State.lean

```diff
@@ -662,3 +662,296 @@ theorem write_mem_irrelevant :
+      have hdiv : i / 8 = 0 :=  Nat.div_eq_of_lt h₁
+      rw [hdiv]
+      simp only [BitVec.add_zero]
+      have hmod : i % 8 = i := Nat.mod_eq_of_lt h₁
+      simp [hmod]
+    · simp only [h₁, decide_False, cond_false]
+      rw [ih]
+      by_cases h₂ : i - 8 < n' * 8
+      · simp only [h₂, decide_True, Bool.true_and, show (i < n' * 8 + 8) by omega]
+        have hi' : ∃ i', i = i' + 8 := by
+          apply Classical.byContradiction
+          intros h
+          simp only [not_exists] at h
+          specialize h (i - 8)
+          omega
+        obtain ⟨i', hi'⟩ := hi'
+        subst hi'
+        simp only [Nat.add_sub_cancel, Nat.zero_lt_succ, Nat.add_div_right, Nat.add_mod_right]
+        congr 2
+        rw [BitVec.add_assoc]
+        congr
+        rw [BitVec.add_def]
+        congr 1
+        simp only [BitVec.toNat_ofNat, Nat.reducePow, Nat.reduceMod]
+        rw [Nat.mod_eq_of_lt]
+        · omega
+        · omega
+      · simp only [h₂, decide_False, Bool.false_and, Bool.false_eq, Bool.and_eq_false_imp,
+          decide_eq_true_eq]
+        intros h₃
+        omega
+      · omega
+
+/--
+A variant of `write_mem` that directly talks about writes to memory, instead of over the entire `ArmState`,
+-/
+def write_mem' (addr : BitVec 64) (val : BitVec 8) (m : Memory) : Memory :=
+  write_store addr val m
+
+theorem write_mem'_of_eq (hix : ix = addr) : write_mem' addr val s ix = val := by
+  simp only [write_mem']
+  subst ix
+  apply store_read_over_write_same
+
+theorem write_mem'_of_neq (hix : ix ≠ addr) : write_mem' addr val s ix = s ix := by
+  simp only [write_mem']
+  apply store_read_over_write_different
+  assumption
+
+theorem write_mem_eq_write_mem' :  (write_mem addr val s).mem = write_mem' addr val s.mem := rfl
+
+def write_mem_bytes' (n : Nat) (addr : BitVec 64)
+    (val : BitVec (n * 8)) (s : Memory) : Memory :=
+  match n with
+  | 0 => s
+  | n' + 1 =>
+    let byte := BitVec.extractLsb 7 0 val
+    let s := write_mem' addr byte s
+    let val_rest := BitVec.zeroExtend (n' * 8) (val >>> 8)
+    write_mem_bytes' n' (addr + 1#64) val_rest s
+
+@[simp]
+theorem write_mem_bytes_eq_write_mem_bytes' (s : ArmState) :
+    write_mem_bytes n addr val s =
+    { s with mem := write_mem_bytes' n addr val s.mem } := by
+  induction n generalizing addr s
+  case zero => simp [write_mem_bytes, write_mem_bytes']
+  case succ n' ih =>
+    simp [write_mem_bytes, write_mem_bytes', read_mem_eq_read_mem', ih,
+      write_mem, write_mem']
+
+/-- Writing zero bytes does not change memory. -/
+theorem write_mem_bytes'_zero : write_mem_bytes' 0 addr val s = s := rfl
+
+/-- Writing (n + 1) bytes can be described as writing `n` bytes and then recursing to write the rest. -/
+theorem write_mem_bytes'_succ :
+    write_mem_bytes' (n + 1) addr val s =
+    let byte := BitVec.extractLsb 7 0 val
+    let s := write_mem' addr byte s
+    let val_rest := BitVec.zeroExtend (n * 8) (val >>> 8) -- TODO: rewrite this as 'truncate'.
```

Comment:
Is this a "pre-merge" TODO, or a "follow up PR" TODO?

Reply:
This is a "follow up" TODO.

Reply:
Forgive my ignorance: why is `truncate` preferred over `zeroExtend`?

Reply:
Aha, it felt like the correct operation, but I realize that I was thinking about it wrong, and that we are in fact zero extending. Happy to remove the TODO.